File: src/Catty/Parser&Serialializer/New Parser & Serializer/XMLHandler/Bricks/Look/SetLookBrick+CBXMLHandler.m
/**
 *  Copyright (C) 2010-2016 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import "SetLookBrick+CBXMLHandler.h"
#import "CBXMLValidator.h"
#import "GDataXMLElement+CustomExtensions.h"
#import "Look+CBXMLHandler.h"
#import "CBXMLParserContext.h"
#import "CBXMLSerializerContext.h"
#import "CBXMLParserHelper.h"
#import "CBXMLSerializerHelper.h"

@implementation SetLookBrick (CBXMLHandler)

+ (instancetype)parseFromElement:(GDataXMLElement*)xmlElement withContext:(CBXMLParserContext*)context
{
    SetLookBrick *setLookBrick = [self new];
    if([xmlElement childCount] == 0) {
        return setLookBrick;
    }
    
    [CBXMLParserHelper validateXMLElement:xmlElement forNumberOfChildNodes:1];
    
    GDataXMLElement *lookElement = [[xmlElement children] firstObject];
    NSMutableArray *lookList = context.spriteObject.lookList;

    Look *look = nil;
    if ([CBXMLParserHelper isReferenceElement:lookElement]) {
        GDataXMLNode *referenceAttribute = [lookElement attributeForName:@"reference"];
        NSString *xPath = [referenceAttribute stringValue];
        lookElement = [lookElement singleNodeForCatrobatXPath:xPath];
        [XMLError exceptionIfNil:lookElement message:@"Invalid reference in SetLookBrick. No or too many looks found!"];
        GDataXMLNode *nameAttribute = [lookElement attributeForName:@"name"];
        [XMLError exceptionIfNil:nameAttribute message:@"Look element does not contain a name attribute!"];
        look = [CBXMLParserHelper findLookInArray:lookList withName:[nameAttribute stringValue]];
        [XMLError exceptionIfNil:look message:@"Fatal error: no look found in list, but should already exist!"];
    } else {
        // OMG!! a look has been defined within the brick element...
        look = [context parseFromElement:xmlElement withClass:[Look class]];
        [XMLError exceptionIfNil:look message:@"Unable to parse look..."];
        [lookList addObject:look];
    }
    setLookBrick.look = look;
    return setLookBrick;
}

- (GDataXMLElement*)xmlElementWithContext:(CBXMLSerializerContext*)context
{
    NSUInteger indexOfBrick = [CBXMLSerializerHelper indexOfElement:self inArray:context.brickList];
    GDataXMLElement *brick = [GDataXMLElement elementWithName:@"brick" xPathIndex:(indexOfBrick+1) context:context];
    [brick addAttribute:[GDataXMLElement attributeWithName:@"type" escapedStringValue:@"SetLookBrick"]];
    if (self.look) {
        if([CBXMLSerializerHelper indexOfElement:self.look inArray:context.spriteObject.lookList] == NSNotFound) {
            self.look = nil;
        } else {
            GDataXMLElement *referenceXMLElement = [GDataXMLElement elementWithName:@"look" context:context];
            NSString *refPath = [CBXMLSerializerHelper relativeXPathToLook:self.look inLookList:context.spriteObject.lookList];
            [referenceXMLElement addAttribute:[GDataXMLElement attributeWithName:@"reference" escapedStringValue:refPath]];
            [brick addChild:referenceXMLElement context:context];
        }
    }
    return brick;
}

@end

File: src/Catty/XML/XMLHandler/Bricks/LookBricks/SetLookBrick+CBXMLHandler.swift
/**
 *  Copyright (C) 2010-2021 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

extension SetLookBrick: CBXMLNodeProtocol {
    static func parse(from xmlElement: GDataXMLElement, with context: CBXMLParserContext) -> Self {

        let setLookBrick = self.init()
        let children = xmlElement.childrenWithoutCommentsAndCommentedOutTag() as Array
        if children.isEmpty {
            return setLookBrick
        }

        CBXMLParserHelper.validate(xmlElement, forNumberOfChildNodes: 1)

        let lookElement = children.first as? GDataXMLElement
        let lookList = context.spriteObject.lookList

        if CBXMLParserHelper.isReferenceElement(lookElement) {
            let referenceAttribute = lookElement?.attribute(forName: "reference")
            let xPath = referenceAttribute?.stringValue()
            guard let element = lookElement?.singleNode(forCatrobatXPath: xPath) else {
                fatalError("Invalid reference in SetLookBrick. No or too many looks found!")
            }
            guard let nameAttribute = element.attribute(forName: "name") else {
                fatalError("Look element does not contain a name attribute!")
            }
            setLookBrick.look = CBXMLParserHelper.findLook(in: lookList as? [Any], withName: nameAttribute.stringValue())
        } else {
            guard let look = context.parse(from: xmlElement, withClass: Look.self as? CBXMLNodeProtocol.Type) as? Look else {
                    fatalError("Unable to parse look...")
                }
            lookList?.add(look)
            setLookBrick.look = look
        }
        return setLookBrick
    }

    func xmlElement(with context: CBXMLSerializerContext) -> GDataXMLElement? {
        let brick = super.xmlElement(for: "SetLookBrick", with: context)

        if self.look != nil {
            if CBXMLSerializerHelper.index(ofElement: self.look, in: context.spriteObject.lookList as? [Any]) == NSNotFound {
                self.look = nil
            } else {
                let referenceXMLElement = GDataXMLElement.init(name: "look", context: context)
                let depthOfResource = CBXMLSerializerHelper.getDepthOfResource(self, for: context.spriteObject)
                let refPath = CBXMLSerializerHelper.relativeXPath(to: self.look, inLookList: context.spriteObject.lookList as? [Any], withDepth: depthOfResource)

                referenceXMLElement?.addAttribute(GDataXMLElement.attribute(withName: "reference", escapedStringValue: refPath) as? GDataXMLNode)
                brick?.addChild(referenceXMLElement, context: context)
            }
        }
        return brick
    }
}
